lib/transaction/trace/exclusive-time-calculator.js

```diff
@@ -7,40 +7,32 @@
 
 class ExclusiveCalculator {
   constructor(root, trace) {
-    this.trace = trace
-    this.id = root.id
-    this.toProcess = [root]
+    this.node = trace.getNode(root.id)
     // use a second stack to do a post-order traversal
     this.parentStack = []
   }
 
   /**
    * Kicks off the exclusive duration calculation.  This is performed
-   * using a depth first, postorder traversal over the tree.
+   * using a depth first, postorder traversal over the tree recursively
+   *
+   * @param {Node} node to process duration and children
    */
-  process() {
-    while (this.toProcess.length) {
-      const segment = this.toProcess.pop()
-      const children = this.trace.getChildren(segment.id)
-      // when we hit a leaf, calc the exclusive time and report the time
-      // range to the parent
-      if (children.length === 0) {
-        segment._exclusiveDuration = segment.getDurationInMillis()
-        if (this.parentStack.length) {
-          this.finishLeaf(segment.timer.toRange())
-        }
-      } else {
-        // in the case we are processing an internal node, we just push it on the stack
-        // and push its children to be processed. all processing will be done after its
-        // children are all done (i.e. postorder)
-        this.parentStack.push({
-          childrenLeft: children.length,
-          segment: segment,
-          childPairs: []
-        })
-        for (let i = children.length - 1; i >= 0; --i) {
-          this.toProcess.push(children[i])
-        }
+  process(node = this.node) {
+    const { children, segment } = node
+    if (children.length === 0) {
+      segment._exclusiveDuration = segment.getDurationInMillis()
+      if (this.parentStack.length) {
```

Comment:
Is this meant to finalize the last node in the tree? Wouldn't that node the be the zeroth element?

Reply:
honestly i'm not really sure. I was hoping to use #2695 and change how we calculate this exclusive time. This seems overly complicated

Reply:
👍